docs: docs for pop bench command (#25)
r0gue-io/pop-docs: pop-cli-for-appchains/guides/benchmarking-a-parachain/benchmarking-pallets-and-extrinsics.md

│  ○ (1) - Extrinsics: 4 selected
│  ○ (2) - Runtime path: tests/runtimes/base_parachain_benchmark.wasm
│  ○ (3) - Genesis builder: runtime
│  ○ (4) - Genesis builder preset: development
│  ○ (5) - Steps: 50
│  ○ (6) - Repeats: 20
│  ○ (7) - High: None
│  ○ (8) - Low: None
│  ○ (9) - Map size: 1000000
│  ○ (10) - Database cache size: 1024
│  ○ (11) - Additional trie layer: 2
│  ○ (12) - No median slope: false
│  ○ (13) - No min square: false
│  ○ (14) - No storage info: false
│  ○ (15) - Weight file template: None
│  ● > Save all parameter changes and continue
└
```

For example, to update the `High` parameter:

```bash
◇  Select the parameter to update:
│  (7) - High: None
│
◆  Provide range values to the parameter "High" (numbers separated by commas)
│  10,50,100
└
```

And you will see the value is updated in the menu:

```bash
│  ○ (6) - Repeats: 20
│  ● (7) - High: 10,50,100
│  ○ (8) - Low: None
```

**Save weight output and parameters to file**

```bash
◆  Provide the output file path for benchmark results (optional).
│  ./weight.rs
└
```
If not provided, no weight output will be saved.

```bash
◆  Provide the output path for benchmark parameters

> │  pop-bench.toml

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/8169db4d544a949697655a91ddcc1a3a4bc75061

└
```

To save parameters, the file path is default to `pop-bench.toml`. With the saved parameter file, you can load parameters from the file by using the `-f` or `--bench-file` flag.

```bash
pop bench -f pop-bench.toml
```

For more advanced parameters configuration, you can run the following command to learn more:

```bash
pop bench --help
```

#### Learning Resources

> * 🧑‍🏫 To learn about Polkadot in general, [Polkadot.network](https://polkadot.network/) website is a good starting point.

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/8169db4d544a949697655a91ddcc1a3a4bc75061

```bash
pop bench pallet --no-build
```

**Configure the genesis builder policy and preset**

```bash
◆  Select the genesis builder policy:
│  ● none (Do not provide any genesis state.)
│  ○ runtime
└
```

Genesis builder policy defines the way to construct the [initial genesis state](https://docs.polkadot.com/develop/parachains/deployment/generate-chain-specs/). There are two options:

- `none`: Do not provide any genesis state.
- `runtime`: Use the runtime's genesis preset.

To configure the genesis builder manually, you can use the `--genesis-builder` flag. For example:

```bash
pop bench pallet --runtime=target/release/pop-runtime-devnet.wasm --genesis-builder=runtime
```

The genesis preset is configured on the runtime via [`sp_genesis_builder`](https://docs.rs/sp-genesis-builder/latest/sp_genesis_builder/) crates which contains a runtime-api to be implemented by runtimes, in order to express their genesis state.

```bash
◇  Select the genesis builder policy:
│  runtime
│
◇  Found 2 genesis builder presets
│
◆  Select the genesis builder preset:
│  ● development
│  ○ local_testnet
└
```

> To configure the genesis builder preset, you can use the `--genesis-builder-preset` flag:

Reviewer: ```suggestion
To configure the genesis builder preset manually, you can use the `--genesis-builder-preset` flag:
```

Please ensure all statements that show the parameter usage include 'manually' as above.

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

```bash
pop bench pallet --runtime=target/release/pop-runtime-devnet.wasm --genesis-builder=runtime --genesis-builder-preset=development
```

**Select pallets and extrinsics**

> You'll be prompted to benchmark all pallets or select a specific one. Pop CLI lists available pallets, allowing you to search by name and choose from the list.

Reviewer: ```suggestion
You'll be prompted to benchmark all pallets or select a specific one. Pop CLI lists all available pallets within the runtime, allowing you to search by name and choose from the list.
```

Author: REsolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

```bash
◆  🔎 Search for a pallet to benchmark
│
│  ● cumulus_pallet_parachain_system
│  ○ cumulus_pallet_xcmp_queue
│  ○ frame_system
│  ○ pallet_balances
│  ○ pallet_collator_selection
│  ○ pallet_message_queue
│  ○ pallet_session
│  ○ pallet_sudo
│  ○ pallet_timestamp
└
```

You can simply search for pallets by typing their name and selecting them from the list.
```bash
◆  🔎 Search for a pallet to benchmark
│  pallet_time
│  ● pallet_timestamp
│  ○ pallet_session
│  ○ pallet_sudo
│  ○ pallet_message_queue
│  ○ pallet_balances
│  ○ pallet_collator_selection
└
```

> If a pallet is specified, the CLI will prompt you to select the extrinsics you want to benchmark within that pallet. You can select multiple extrinsics by pressing the spacebar. Same as searching for pallets, you can search for extrinsics by typing their name as well.

Reviewer: We seem to use extrinsics and dispatchables interchangeably. I know the underlying tooling uses extrinsic, as you are technically submitting extrinsics (transactions) to measure their performance, but I think we should use 'dispatchables' or 'dispatchable functions' when referring to the functions by name. 'functions' would also be fine, as in pallet functions.

```bash
◆  🔎 Search for extrinsics to benchmark (select with space)
│
│  ◼ burn_allow_death
│  ◼ burn_keep_alive
│  ◼ force_adjust_total_issuance
│  ◼ force_set_balance_creating
│  ◻ force_set_balance_killing
│  ◻ force_transfer
│  ◻ force_unreserve
│  ◻ transfer_all
│  ◻ transfer_allow_death
│  ◻ transfer_keep_alive
│  ◻ upgrade_accounts
└
```

> If `--extrinsic=` and `--pallet=` are provided, the CLI will skip the search and directly benchmark with the specified arguments.

Reviewer: This counters my point above. 🙃

> **Manage parameters using the menu**

Reviewer: ```suggestion
**Manage parameters**
```

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

> Pop CLI displays all editable parameters in a menu, allowing you to preview and modify them as needed. Simply select a parameter to update.

Reviewer: ```suggestion
Pop CLI displays all editable parameters, allowing you to preview and modify them as needed. Simply select a parameter to update.
```

We dont need to say menu, its obvious.

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

> > To skip the parameter menu, you can use the `--skip-menu` flag.

Reviewer: ```suggestion
> To skip parameter configuration, you can use the `--skip-parameters` flag.
```

I know this will require in the underlying code, but I feel this is better than using 'menu'. Menu is just a UI presentation, not something specific to the core domain. In this case that is 'providing benchmarking parameter values'.

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

```bash
◆  Select the parameter to update:
│  ○ (0) - Pallets: pallet_balances
│  ○ (1) - Extrinsics: 4 selected
│  ○ (2) - Runtime path: tests/runtimes/base_parachain_benchmark.wasm
│  ○ (3) - Genesis builder: runtime
│  ○ (4) - Genesis builder preset: development
│  ○ (5) - Steps: 50
│  ○ (6) - Repeats: 20
│  ○ (7) - High: None
│  ○ (8) - Low: None
│  ○ (9) - Map size: 1000000
│  ○ (10) - Database cache size: 1024
│  ○ (11) - Additional trie layer: 2
│  ○ (12) - No median slope: false
│  ○ (13) - No min square: false
│  ○ (14) - No storage info: false
│  ○ (15) - Weight file template: None
│  ● > Save all parameter changes and continue
└
```

For example, to update the `High` parameter:

```bash
◇  Select the parameter to update:
│  (7) - High: None
│
◆  Provide range values to the parameter "High" (numbers separated by commas)
│  10,50,100
└
```

And you will see the value is updated in the menu:

```bash
│  ○ (6) - Repeats: 20
│  ● (7) - High: 10,50,100
│  ○ (8) - Low: None
```

> **Save weight output and parameters to file**

Reviewer: ```suggestion
**Save weight output and parameter values to file**
```

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

> After the parameter menu, you will be prompted to save the weight output and parameters to a file. If not provided, no weight output will be saved.

Reviewer: ```suggestion
After the parameter menu, you will be prompted to save the weight output and provided parameter values to a file. If not provided, no weight output will be saved.
```

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

```bash
◆  Provide the output file path for benchmark results (optional).
│  ./weight.rs
└
```

> To save parameters, the file path is default to `pop-bench.toml`.

Reviewer: ```suggestion
To save parameter values, the file path is default to `pop-bench.toml`.
```

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

```bash

> ◆  Provide the output path for benchmark parameters

Reviewer: ```suggestion
◆  Provide the output path for benchmark parameter values
```

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

│  pop-bench.toml
└
```

Below is the example of the parameter file:

```toml
version = "1"
pallet = "pallet_timestamp"
extrinsic = "*"
exclude_pallets = []
all = false
steps = 50
lowest_range_values = []
highest_range_values = []
repeat = 20
external_repeat = 1
json_output = false
no_median_slopes = false
no_min_squares = false
output_pov_analysis = "median-slopes"
no_verify = false
extra = false
runtime = "runtimes/base_parachain_benchmark.wasm"
allow_missing_host_functions = false
genesis_builder = "Runtime"
genesis_builder_preset = "development"
database_cache_size = 1024
list = false
no_storage_info = false
worst_case_map_values = 1000000
additional_trie_layers = 2
disable_proof_recording = false
skip_menu = false
skip_confirm = false
```

> With the saved parameter file, you can load parameters from the file by using the `-f` or `--bench-file` flag.

Reviewer: ```suggestion
With the saved parameter file, you can load parameter values from the file by using the `-f` or `--bench-file` flag.
```

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b

```bash
pop bench -f pop-bench.toml
```

> For more advanced parameters configuration, you can run the following command to learn more:

Reviewer: ```suggestion
For more advanced parameter configuration, you can run the following command to learn more:
```

Author: Resolved in https://github.com/r0gue-io/pop-docs/pull/25/commits/1e2c6f371406466c64a93093c37aef9bc40bdd1b
